tpu-provisioner/internal/controller/synced_start_controller.go

```diff
@@ -0,0 +1,223 @@
+	lg := ctrllog.FromContext(ctx)
+
+	lg.V(3).Info("Reconciling JobSet")
+
+	js := &jobset.JobSet{}
+	if err := r.Get(ctx, req.NamespacedName, js); err != nil {
+		if apierrors.IsNotFound(err) {
+			// Don't requeue, no longer exists.
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, fmt.Errorf("getting jobset: %w", err)
+	}
+
+	if js.GetDeletionTimestamp() != nil {
+		return ctrl.Result{}, nil
+	}
+
+	if js.Annotations == nil {
+		return ctrl.Result{}, nil
+	}
+	configMapName := js.Annotations[jobsetSyncedStartAnnotation]
+	if configMapName == "" {
+		return ctrl.Result{}, nil
+	}
+
+	// NOTE: If the ConfigMap is set as an optional volume in the Pod spec
+	// the propagation of a ConfigMap creation into container mount
+	// appears to take a long time (>30s in a test).
+	cm := &corev1.ConfigMap{}
+	if err := r.Get(ctx, types.NamespacedName{
+		Namespace: req.Namespace,
+		Name:      configMapName,
+	}, cm); err != nil {
+		if apierrors.IsNotFound(err) {
+			// Create ConfigMap if it did not exist.
+			cm = &corev1.ConfigMap{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      configMapName,
+					Namespace: req.Namespace,
+				},
+			}
+			if err := ctrl.SetControllerReference(js, cm, r.Scheme); err != nil {
+				return ctrl.Result{}, fmt.Errorf("setting owner reference: %w", err)
+			}
+			if err := r.Create(ctx, &corev1.ConfigMap{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      configMapName,
+					Namespace: req.Namespace,
+				},
+			}); err != nil {
+				if apierrors.IsAlreadyExists(err) {
+					return ctrl.Result{Requeue: true}, nil
+				}
+				return ctrl.Result{}, fmt.Errorf("creating ConfigMap: %w", err)
+			}
+		}
+		return ctrl.Result{}, fmt.Errorf("getting ConfigMap: %w", err)
+	}
+
+	if cm.Data == nil {
+		cm.Data = map[string]string{}
+	}
+
+	// If the ConfigMap has already been updated, return early.
+	if cm.Data[allStartedConfigMapKey] == "true" {
+		return ctrl.Result{}, nil
+	}
+
+	// List all Pods in the current restart of the JobSet.
+	var pods corev1.PodList
+	if err := r.List(ctx, &pods, client.MatchingLabels{
+		jobsetNameLabel:    js.Name,
+		jobsetRestartLabel: strconv.Itoa(int(js.Status.Restarts)),
+	}); err != nil {
+		return ctrl.Result{}, fmt.Errorf("listing pods: %w", err)
+	}
+
+	threshold, ok := startedThreshold(js)
+	if !ok {
+		return ctrl.Result{}, nil
+	}
+
+	// TODO: See if it is possible to add an index to .status.startTime
+	// to avoid the for loop.
+	var startedCount int32
+	var thresholdMet bool
+	for _, p := range pods.Items {
+		if p.Status.StartTime != nil {
+			startedCount++
+			if startedCount >= threshold {
+				thresholdMet = true
+				break
+			}
+		}
+	}
+
+	if !thresholdMet {
+		return ctrl.Result{}, nil
+	}
+
+	// Update the ConfigMap to indicate that all Pods have started.
+	cm.Data[allStartedConfigMapKey] = "true"
+
+	if err := ctrl.SetControllerReference(js, cm, r.Scheme); err != nil {
```

Comment:
I think we are setting the configMap controller reference twice accidentally (first time on line 103)

Reply:
The thought here is that the user might have created the configmap and the controller would adopt it. So I set the reference before an update in case the create-path was never taken. It confused me re-reading it too, might try to refactor to something cleaner.

Reply:
I see. If the configMap already has the owner reference, will this line return an error and prevent the Update() call below from being executed?

Reply:
I removed this line: I think if the user created the CM they are responsible for deleting it. If the controller created it - it will assign the JobSet as the owner. Thoughts?

Reply:
Sounds good to me

Reply:
One more thing, I notice the configMap name is hardcoded, so we are using the same configMap for all jobsets? Shouldn't we create 1 configMap per JobSet?